File: Java/Anagrams.java
/*
Two strings, s1  and s2, are called anagrams if they contain all the same characters in the same frequencies. 

For example, the anagrams of CAT are CAT, ACT, TAC, TCA, ATC, and CTA.
*/
import java.util.*;
class Anagrams
{
    public static void main(String[] args) {
        
        Scanner sc = new Scanner(System.in);

        // Taking input from the user
        String s1 = sc.nextLine();
        String s2 = sc.nextLine();

        //Storing the result and printing it.
        boolean result = isAnagram(s1, s2);
        System.out.println(result);

    }
    public static boolean isAnagram(String a,String b)
    {
        //Auxillary array to store the count of the characters present in the strings
        int arr[] = new int[256];

        //Coverting both the strings to character arrays to store the count of each character in the auxillary array.
        char a1[] = a.toCharArray();
        char b1[] = b.toCharArray();

        // To store the sum of the elements of the auxillary array
        int c = 0;

        /* The auxillary array indexes actually represents the ASCII codes of the characters for all the characters in the Strings.
            If any character is found, then the value of the character index position in the auxillary array increases by 1 every time
            so it is first passed through the first string(i.e where value increases) and then passed through the second string(i.e where
            value decreses) and all the elements of the auxillary array becomes 0.*/
        for(int i=0;i<a1.length;i++)
        {
            arr[a1[i]]++;
        }
        for(int i=0;i<b1.length;i++)
        {
            arr[b1[i]]--;
        }

        for(int i=0;i<256;i++)
        {
            c += Math.abs(arr[i]);
        }
        
        // Checking whether the values of the auxillary array are null or not.
        if(c==0)
            return true;
        else
            return false;
    }
}
/*
Input:

The first line contains a lowercase string denoting s1.
The second line contains a lowercase string denoting s2.

Output:

Print "true" if s1 and s2 are case-insensitive anagrams of each other; otherwise, print "false" instead.

Example 1:

Input 1:

anagram
margana

Output 1:

true

Input 2:

anagramm
marganaa

Output 2:

false

Time Complexity : O(N + M), N and M are the lengths of Strings s1 & s2 respectively.
Space Complexity : O(N + M), N and M are the lengths of Strings s1 & s2 respectively.

*/

File: C-Plus-Plus/anagrams.cpp
// Check if two strings are anagrams or not

/*
Two strings, a and b, are called anagrams if they contain all the same characters in the same frequencies.
For example, the anagrams of CAT are CAT, ACT, TAC, TCA, ATC, and CTA.
*/

#include<bits/stdc++.h>
using namespace std;

bool check_anagram(string s1, string s2)
{
    //If the size of both strings is not same, frequency of each character can never be same
    if(s1.size()!=s2.size())
        return false;

    map<char,int> m1;

    //Increasing the frequency of character that appeared in s1 by 1
    for(int i=0; i<s1.size(); i++)
        m1[s1[i]]++;

    //Decreasing the frequency of character that appeared in s2 by 1
    for(int i=0; i<s2.size(); i++)
        m1[s2[i]]--;

    //Checking if any frequency is non-zero
    for(auto it=m1.begin(); it!=m1.end(); it++)
    {
        if(it->second!=0)
            return false;
    }

    return true;
}

int main()
{
    string s1,s2;

    //Taking input from user
    cout<<"Enter String 1: ";
    cin>>s1;

    cout<<"Enter String 2: ";
    cin>>s2;

    //Using check_anagram function to know if both strings are anagrams or not
    if(check_anagram(s1,s2))
        cout<<"The Strings "<<s1<<" and "<<s2<<" are anagrams!";
    else
        cout<<"The Strings "<<s1<<" and "<<s2<<" are not anagrams!";

    return 0;
}

/*
Sample Input 1:
aabbbccc
abcabcabc
Sample Output 1:
The Strings aabbbccc and abcabcabc are not anagrams!
Sample Input 2:
aabbbccc
abcabcbc
Sample Output 1:
The Strings aabbbccc and abcabcbc are anagrams!
*/

/*
Time Complexity: O(s1.size())
Space Complexity: O(s1.size())
*/
